test/integration/cluster_filter_integration_test.cc

```diff
@@ -17,7 +17,7 @@ class PoliteFilter : public Network::Filter, Logger::Loggable<Logger::Id::filter
   Network::FilterStatus onData(Buffer::Instance& data, bool end_stream) override {
     ENVOY_CONN_LOG(debug, "polite: onData {} bytes {} end_stream", read_callbacks_->connection(),
                    data.length(), end_stream);
-    if (!read_greeted_) {
+    if (!read_greeted_ && on_new_connection_called_ == 1) {
```

Comment:
nit: maybe put ASSERT in the destructor to assert 0 or 1?

Reply:
Changed to test with and without this change via runtime guard `initialize_upstream_filters` and passing info from test filter to the test class to be explicitly asserted in the right places of the test body.